HeatmapExercises

jkc9886 opened the pull request:
1. Added new exercises (Basic heatmap ex-5)
2. Extra: Interactive heatmap & custom color palette 
3. Rewrote the k-means clustering and hierarchial exercises  ( #518  Faced problems in using sechm and addcluster(), the BLUSPARAM parameter and sechm features )

jkc9886 committed:
Exercises.qmd

1. Added new exercises (Basic heatmap ex-5)
2. Extra: Interactive heatmap & custom color palette 
3. Rewrote the k-means clustering and hierarchial exercises  ( #518  Faced problems in using sechm and addcluster(), the BLUSPARAM parameter and sechm features )

jkc9886 commented:
@TuomasBorman 

I opened this new PR and resolved the previous one( #518) because the file was not updated with recent changes and I was having trouble in knitting it. I can give using sechm plot more tries but I might need your guidance there, sechm does not process on my version( I downloaded the latest one though) so I am really confused and stuck there. I wrote the k-means clustering and hierarchial codes again using Complexheatmap as we were discussing in the comments earlier and these build well locally. Please give me with your valuable suggestions :)

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,134 @@ p <- Heatmap(
-# Assuming enterotype is a SummarizedExperiment object with sample metadata
-# and assay_data is a matrix containing assay data
+# Perform hierarchical clustering on the MDS data (extracted in last exercise)
+dist_matrix <- dist(mds_data)  # Compute distance matrix on MDS dimensions
+hclust_result <- hclust(dist_matrix, method = "complete")
+
+# Cut the dendrogram into 3 clusters
+cluster_labels <- cutree(hclust_result, k = 3)
+
+# Add cluster labels to the SummarizedExperiment object
+colData(tse)$cluster <- factor(cluster_labels)
+
```
We have mia::addCluster that can perform these operations with a single line. Could you switch to that?

jkc9886 replied:
addCluster() is generally used with predefined clustering methods in the bluster package, such as KmeansParam(I switched the previous exercise to using addCluster) but there is no predefined parameter for hierarchical clustering, so addCluster does not directly fit this usage pattern. It is rather complicating it since I am having to define a new dummy BLUSPARAM, do you think it will be too complex that way?

antagomir replied:
Ah, this is interesting. 

Quick search shows that there are ready-made solutions for hierarchical clustering of SE objects:
1) [scran::quickCluster()](https://rdrr.io/bioc/scran/man/quickCluster.html)
2) [tomoda::hierarchClust](https://s3.jcloud.sjtu.edu.cn/899a892efef34b1b944a19981040f55b-oss01/bioconductor/3.14/bioc/manuals/tomoda/man/tomoda.pdf)
3) [clusterExperiment](https://www.bioconductor.org/packages/release/bioc/vignettes/clusterExperiment/inst/doc/clusterExperimentTutorial.html#plotHeatmap)

Perhaps some other solutions exists as well.

Could you check if some of these could provide a handy solution?

jkc9886 replied:
Yes sure. It will be interesting to try new ways :))

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,134 @@ p <- Heatmap(
+  labs(title = "MDS with Hierarchical Clustering") +
+  theme_minimal()
+
+# Optional: Visualize the hierarchical clustering with a heatmap
+# Convert MDS data to a matrix for heatmap visualization
+mds_matrix <- t(as.matrix(mds_data))  # Transpose the matrix to match the columns
+
+# Create a heatmap with cluster annotations
+heatmap <- Heatmap(mds_matrix,
+                   name = "MDS",
+                   show_row_dend = TRUE,
+                   show_column_dend = TRUE,
+                   column_split = cluster_labels,
+                   column_title = "Clusters",
+                   top_annotation = HeatmapAnnotation(cluster = cluster_labels),
+                   clustering_distance_columns = dist_matrix,
+                   clustering_method_columns = "complete")
+
+# Draw the heatmap
+draw(heatmap, heatmap_legend_side = "right")
```
Perhaps you can get mds_matrix directly from mia::addCluster output? 

In general we support sechm for heatmaps because it supports the SE tool ecosystem, is there an advantage from using Heatmap.

jkc9886 replied:
I had a talk with @ElySeraidarian, I appreciate his help. I am trying to do it with addCluster and sechm now.

antagomir commented:
You could check with @ElySeraidarian if you need support? He is familiar with ordination visualization methods.

jkc9886 commented:
> You could check with @ElySeraidarian if you need support? He is familiar with ordination visualization methods.

yes that would be really helpful

jkc9886 committed:
Update exercises.qmd

Used the addCluster() function in the k-means clustering visualization exercise.

jkc9886 committed:
Updating extra-hierarchial_clustering_heatmap_exercise

To make it a simpler code : 
1.  I loaded random gene and sample data, divided the samples into three clusters, and added this cluster information to the metadata. 
2. Then combined the expression data and metadata into a SummarizedExperiment (se) object, and,
3. Used sechm function to create a heatmap, clustering both rows (genes) and columns (samples).

antagomir commented:
Can you resolve the conflicts and also resolve conversations that you have already addressed fully?

jkc9886 commented:
This PR is ready to be merged.

jkc9886 committed:
Merge branch 'devel' into jkc9886-patch-6

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1552,6 +1552,24 @@ sechm(
+library(pheatmap)
+
+# Create example data
+set.seed(123)
+example_data <- matrix(rnorm(200), nrow = 20, ncol = 10)
+
+# Create basic heatmap using pheatmap
+pheatmap(example_data)
```
We should use the ready made demo data sets from real studies (included in mia, miaViz, miaTime packages).

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
+library(heatmaply)
+
+# Create example data
+example_data <- matrix(rnorm(200), nrow = 20, ncol = 10)
+
+# Create interactive heatmap
+heatmaply(example_data)
```
use real example data

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
-pca_aitchison_clusters <- kmeans(reducedDim(pca_aitchison), centers = 3)$cluster
+# Perform k-means clustering (k = 3) using the base R kmeans function
+set.seed(123)
+kmeans_result <- kmeans(mds_data, centers = 3)
```
We should use the mia clustering functions instead of base R. This should require less code.

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
-# and assay_data is a matrix containing assay data
+# Generate sample abundance data
+set.seed(123)
+sample_data <- matrix(rnorm(100), nrow = 10)
+rownames(sample_data) <- paste0("Gene", 1:10)
```
use real demo data instead

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
+colnames(sample_data) <- paste0("Sample", 1:10)
+
+# Perform hierarchical clustering
+hclust_result <- hclust(dist(t(sample_data)), method = "complete")
```
use mia clustering functions instead of base R

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
+rownames(metadata) <- colnames(sample_data)
+
+# Create a SummarizedExperiment object
+se <- SummarizedExperiment(assays = list(counts = sample_data),
+                           colData = metadata)
```
use ready made real demo data instead

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
+#| eval: false
+
+# Load necessary libraries
+library(ComplexHeatmap)
```
Should we use sechm or ComplexHeatmap?

jkc9886 replied:
I just wanted to demostrate the use of circlize::colorRamp2 with sechm, the code would be much longer since the sechm function has some default settings that override the color argument or apply additional settings that affect color rendering and we will have to log transform the data for it. that's why I showed the custom color prompt with ComplexHeatmap. Should I swap with sechm?

antagomir replied:
Ok I see. I think sechm is ok.

Why this has "eval: false"? Could we have it evaluated (true)?

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1674,71 +1692,126 @@ p <- Heatmap(
-# Extract sample metadata
-sample_metadata <- colData(enterotype)
+# Create example data
+example_data <- matrix(rnorm(200), nrow = 20, ncol = 10)
```
use real demo data

jkc9886 replied:
Sure!

antagomir reviewed (changes requested):
This PR should take more use of the demo data sets and ready-made mia/miaViz functions because OMA is not book about base R but a book about microbiome data science using the dedicated Bioconductor methods for that purpose.

antagomir committed:
Merge branch 'devel' into jkc9886-patch-6

jkc9886 committed:
Update exercises.qmd

Baes on requested changes in the last commit :
1. Used real demo data(from mia) in each example
2. Used mia clustering functions(addcluster(), bluster - BLUSPARAM) and shortened the mentioned codes.

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1560,11 +1560,14 @@ sechm(
+data(enterotype)
+
+# Extract the data matrix for heatmap
+example_data <- assay(enterotype, "counts")
```
Is it necessary the extract the data matrix, or could we operate on the tse data object directly using sechm?

sechm provides a simple interface between SummarizedExperiment objects and the ComplexHeatmap package. It enables plotting annotated heatmaps from SE objects, with easy access to rowData and colData columns, and implements a number of features to make the generation of heatmaps easier and more flexible. 
https://www.bioconductor.org/packages/release/bioc/html/sechm.html

antagomir replied:
This comment applies regardless of ComplexHeatmap or sechm

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1703,17 +1706,27 @@ Create an interactive heatmap.
+
+# Here we extract the first 20 taxa and the first 10 samples as an example
+taxa_data <- assay(GlobalPatterns)[1:20, 1:10]
+taxa_names <- rownames(taxa_data)
+sample_names <- colnames(taxa_data)
```
are these used below (I do not immediately find)? If not, then better to remove lines 1718-1719

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1703,17 +1706,27 @@ Create an interactive heatmap.
-# Create interactive heatmap
-heatmaply(example_data)
+# Create the interactive heatmap using heatmaply
+heatmaply(taxa_data,
```
If heatmaply does not support SE, then 

1) let's consider if this example brings enough added value to be included in OMA (another option is just to mention that it is available but not show an example)

2) if we keep this in OMA, then I would write more directly
`heatmaply(assay(GlobalPatterns)[1:20, 1:10]), ...)`

or even better, aggregate first to higher taxonomic level with agglomerateByRank() so that you can show the entire data instead of a subset.

antagomir replied:
Alternatively, you can check if there is another smaller data set that could be visualized entirely.

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1703,17 +1706,27 @@ Create an interactive heatmap.
+heatmaply(taxa_data,
+          xlab = "Samples",
+          ylab = "Taxa",
+          main = "Interactive Heatmap of
+                  GlobalPatterns Dataset")
```
main is slightly misleading since this is a subset of GlobalPatterns and the entire GlobalPatterns

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1763,36 +1771,35 @@ plotReducedDim(tse, dimred = "MDS") +
-metadata <- data.frame(Cluster = as.factor(clusters))
-rownames(metadata) <- colnames(sample_data)
+# Here we extract the first 20 taxa and the first 20 samples as an example
+sample_data <- assay(GlobalPatterns)[1:20, 1:20]
```
Is this needed?

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1763,36 +1771,35 @@ plotReducedDim(tse, dimred = "MDS") +
 # Create a SummarizedExperiment object
-se <- SummarizedExperiment(assays = list(counts = sample_data),
-                           colData = metadata)
+se <- SummarizedExperiment(assays = list(counts = sample_data))
```
GlobalPatterns is a already a TreeSE object (and also SE object because all TreeSE are SE)

Unless I am mistaken, you are here first extracting "sample_data" (assay) from the GlobalPatterns TreeSE object, and then reconstructing this again as an SE. Is this necessary? Why not use the original GlobalPatterns directly - you can rename it if necessary, with `tse <- GlobalPatterns` as in most examples.

antagomir commented on inst/pages/exercises.qmd:
```diff
@@ -1801,14 +1808,18 @@ Customize the heatmap with a specific color palette to enhance visualization usi
+data("GlobalPatterns", package = "mia")
+
+# Here we extract the first 30 taxa and the first 20 samples as an example
+example_data <- assay(GlobalPatterns)[1:30, 1:20]
```
same issues as above

antagomir commented:
Thanks - it seems that the capabilities of TreeSE could be still accommodated more.

jkc9886 committed:
Shortened codes

1. Used the datasets directly unlike creating the SE object earlier.
2. Substituted larger datasets like GlobalPatterns with smaller datasets like enterotype and esophagus.

antagomir committed:
Merge branch 'devel' into jkc9886-patch-6

jkc9886 committed:
final updates_exercises.qmd

1. removed the assay extraction step for lines 1564-1568
2. Line 1780 (eval : true)

antagomir committed:
Merge branch 'devel' into jkc9886-patch-6

antagomir commented:
@jkc9886 this ready for comments?

TuomasBorman commented on inst/pages/exercises.qmd:
```diff
@@ -1546,12 +1546,28 @@ tse <- transformAssay(tse, assay.type = "counts", method = "relabundance")
+#| eval: false
+
+library(mia)
+library(pheatmap)
```
Why not to use CoomplexHeatmap::pheatmap()

TuomasBorman commented on inst/pages/exercises.qmd:
```diff
@@ -1546,12 +1546,28 @@ tse <- transformAssay(tse, assay.type = "counts", method = "relabundance")
+library(pheatmap)
+
+# Create heatmap using pheatmap on mia's enterotype dataset
+pheatmap(assay(enterotype, "counts"))
```
This is exercise 4 in Heatmap exercises.

In earlier heatmap exercises, "GlobalPatterns" data was used. However, here you use "esophagus". It would be simpler to use same dataset. (at least, the dataset should be loaded in this heatmap section)

TuomasBorman replied:
As this exercise 5 does the same as exercise 4, it should be say this. I think it is little bit confusing why we want to do the same.

-->

It could be justified by saying that we can create the same plot by using the ComplexHeatmap directly

TuomasBorman commented on inst/pages/exercises.qmd:
````diff
@@ -1674,72 +1690,113 @@ p <- Heatmap(
 
-# Setting seed for reproducibility
-set.seed(123)
+# Create the interactive heatmap using aggregated data
+heatmaply(assay(agglomerateByRank(GlobalPatterns, rank = "Phylum")),
+          xlab = "Samples", ylab = "Phyla")
+```
+Notice we aggregate the "GlobalPatterns" dataset
+to a higher taxonomic level to make the data more manageable.
 
````
This example should be structured. Now it is little bit hard to read. Moreover, usually it might be good idea to scale values as now we can only observe those values that are large (other values are just purple).

```
# Agglomerate
# Transform CLR
# Transform Z-score
# Plot heatmap
```

TuomasBorman commented on inst/pages/exercises.qmd:
```diff
@@ -1674,72 +1690,113 @@ p <- Heatmap(
+tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray", assay.type = "counts")
+
+# Perform k-means clustering using mia's addCluster()
+kmeans_param <- KmeansParam(centers = 3)
+tse <- addCluster(tse, colDataName = "kmeans_clusters", BLUSPARAM = kmeans_param)
+
+# Visualize MDS with clusters
+plotReducedDim(tse, dimred = "MDS") +
+  labs(title = "MDS with k-means Clustering") +
+  theme_minimal()
```
Try:

```
# Perform k-means clustering using mia's addCluster()
kmeans_param <- KmeansParam(centers = 3)
tse <- addCluster(tse, by = "cols", clust.col = "kmeans", BLUSPARAM = kmeans_param)

# Visualize MDS with clusters
plotReducedDim(tse, dimred = "MDS", colour_by = "kmeans") +
  labs(title = "MDS with k-means Clustering") +
  theme_minimal()
```

0xMuluh committed:
Merge branch 'devel' into jkc9886-patch-6

0xMuluh committed:
Merge branch 'devel' into jkc9886-patch-6

TuomasBorman committed:
Merge branch 'devel' into jkc9886-patch-6

TuomasBorman committed:
up
